test/test_bigquery.py

```diff
@@ -0,0 +1,235 @@
+from datetime import datetime, timedelta
+from unittest import mock
+
+import pandas as pd
+import pycarol
+
+
+def test_init_bq():
+    ...
+
+
+def test_is_expired1():
+    bq_mock = mock.MagicMock()
+    bq_mock._provided_sa = True
+    assert pycarol.bigquery.BQ.is_expired(bq_mock) is False
+
+
+@mock.patch("pycarol.bigquery.BQ.token", new_callable=mock.PropertyMock)
+def test_is_expired2(mock_token):
+    bq_mock = mock.MagicMock()
+    mock_token.return_value = None
+    assert pycarol.bigquery.BQ.is_expired(bq_mock) is True
+
+
+@mock.patch("pycarol.bigquery.BQ.token", new_callable=mock.PropertyMock)
+def test_is_expired3(mock_token):
+    bq_mock = mock.MagicMock()
+    offset_date = datetime.now() - timedelta(days=1)
+    mock_token.return_value = {"expiration_time": offset_date}
+    assert pycarol.bigquery.BQ.is_expired(bq_mock) is True
+
+
+@mock.patch("pycarol.bigquery.BQ.token", new_callable=mock.PropertyMock)
+def test_is_expired4(mock_token):
+    bq_mock = mock.MagicMock()
+    offset_date = datetime.now() + timedelta(days=1)
+    mock_token.return_value = {"expiration_time": offset_date, "env": {"env_id": 5}}
+    assert pycarol.bigquery.BQ.is_expired(bq_mock) is True
+
+
+@mock.patch("pycarol.bigquery.BQ.token", new_callable=mock.PropertyMock)
+def test_is_expired5(mock_token):
+    bq_mock = mock.MagicMock()
+    offset_date = datetime.now() + timedelta(days=1)
+    mock_token.return_value = {"expiration_time": offset_date, "env": {"env_id": 5}}
+    bq_mock.env = {"env_id": 5}
+    assert pycarol.bigquery.BQ.is_expired(bq_mock) is False
+
+
+@mock.patch("pycarol.bigquery._save_local_cache")
+@mock.patch(
+    "pycarol.bigquery._save_cds_cache",
+)
+@mock.patch("pycarol.bigquery._get_tmp_key", return_value={"sa": "sa"})
+def test_get_credential1(mock_tmpkey, mock_savecds, mock_savelocal):
+    bq_mock = mock.MagicMock()
+    bq_mock.cached_cds = True
+    bq_mock.is_expired = lambda: True
+    ret = pycarol.bigquery.BQ.get_credential(bq_mock)
+    assert ret["sa"] == mock_tmpkey.return_value["sa"]
+    assert "expiration_time" in ret
+    assert mock_savelocal.call_count == 1
+    assert mock_savecds.call_count == 1
+
+
+@mock.patch("pycarol.bigquery.BQ.token", new_callable=mock.PropertyMock)
+def test_get_credential2(mock_token):
+    mock_token.return_value = {"sa": 5}
+    bq_mock = mock.MagicMock()
+    bq_mock.is_expired = lambda: False
+    bq_mock.service_account = None
+    ret = pycarol.bigquery.BQ.get_credential(bq_mock)
+    assert ret == 5
+
+
+def test_get_fetch_cache1():
+    bq_mock = mock.MagicMock()
+    bq_mock.is_expired = lambda: False
+    bq_mock.service_account = 5
+    pycarol.bigquery.BQ.token = 5
+    assert pycarol.bigquery.BQ._fetch_cache(bq_mock) is None
+    assert bq_mock.service_account == 5
```

Comment:
Sugestão:
`assert(bq_mock.service_account == 5, "sa should not be changed in _fetch_cache")`